frontend/ui/elements/log_viewer.lua

```diff
@@ -11,56 +14,109 @@ local function checkFile(f)
     end
 end
 
-local M = {}
+local LogViewer = {}
 
-function M:new(o)
+function LogViewer:new(o)
     o.exists = checkFile(o.file)
     setmetatable(o, self)
     self.__index = self
     return o
 end
 
-function M:truncate()
+function LogViewer:truncate()
     if not self.exists then return end
     UIManager:show(require("ui/widget/confirmbox"):new{
         text = T(_("Really truncate %1?"), self.file),
         ok_callback = function()
             local file = io.open(self.file, "w")
-            if file then file:close() end
-            require("logger").info(T(_("[%1 truncated]"), self.file))
+            local msg = T(_("%1 [%2 truncated]"), os.date("%x-%X"), self.file)
+            if file then
+                file:write(msg .. "\n")
+                file:close()
+            end
         end,
     })
 end
 
-function M:show(maxsize)
-    if not self.exists then return end
+function LogViewer:loadLog()
     local file = io.open(self.file, "r")
-    local tail = type(maxsize) == "number"
+    local tail = type(self.maxsize) == "number"
     if tail then
-        file:seek("end", -maxsize)
+        file:seek("end", -self.maxsize)
     end
     local content = file:read("*all") or _("empty")
     file:close()
     if tail then
-        if content:len() == maxsize then
+        if content:len() == self.maxsize then
             local first_cr = content:find("\n")
             if first_cr then
-                content = content:sub(first_cr+1)
+                content = content:sub(first_cr + 1)
             end
             content = T(_("[Start of %1 not shown]\n"), self.file) .. content
         else
             content = T(_("[Start of %1\n"), self.file) .. content
         end
     end
-    local Font = require("ui/font")
-    local view = require("ui/widget/textviewer"):new{
+    self.content = content
+end
+
+function LogViewer:show(maxsize)
+    if not self.exists then return end
+    self.maxsize = maxsize
+    self:loadLog()
+
+    self.textviewer = TextViewer:new{
         title = self.file,
-        text = content,
+        text = self.content,
         text_face = Font:getFace("infont", 10),
+        width = self.textviewer_width,
+        height = self.textviewer_height,
         justified = false,
+        buttons_table = {
+            {
+                {
+                    text = "<<",
+                    callback = function()
+                        self.textviewer.scroll_text_w:scrollToTop()
+                    end,
+                },
+                {
+                    text = "<",
```

Comment:
Don't forget our nice arrow symbols. ;-)

Reply:
a link would be awesome. I tried to grep sources but didn't spend much time :)

Reply:
https://github.com/koreader/koreader/blob/6b90239e7c6bc72c13efd54cbffc14bbbbce2c77/frontend/apps/reader/modules/readersearch.lua#L147-L150

https://github.com/koreader/koreader/blob/6b90239e7c6bc72c13efd54cbffc14bbbbce2c77/frontend/ui/widget/inputdialog.lua#L744-L753